### ISSUE TITLE
Dispatch based on SID

### ISSUE THREAD

**Issue by film42** [nats-io/nats.java#197]
It would be really nice if `NatsDispatcher` used the underlying `sid` to dispatch instead of a HashMap based on the subscription name. This would allow you to subscribe to the same topic multiple times. Now, this might not make that much sense from the client's point of view, but it does make sense from a "load balance" point of view.

As a sort of "hack" I have our servers subscribe to each topic `n` times, with a several second delay between each round of subscriptions. This allows our servers to warm up slowly. It's also useful for balancing traffic across larger/smaller vms. I'm currently upgrading from the pre-1.0 client, where I could pass a shared queue into the `subscribe` method. I think the dispatcher is really neat way to abstract the problem, but I'm unable to use it, because it doesn't use the underlying sid. 

Would you be open to review a pull request that changes this? I'm guessing it would end up being a breaking change, which is probably undesirable.

Otherwise, would you be interested in extending the connection interface to support:
```
public Subscription subscribe(String subject, String queueName, MessageQueue messageQueue);
```

...so I can support similar behavior to the pre-1.0 client?

**Comment by film42** [comment]
Here's an example of how the `Connection` interface could be extended to support a user provided message queue.
```java
public interface Connection extends AutoCloseable {
    public Subscription subscribe(String subject);
    public Subscription subscribe(String subject, String queueName);
    // Expose a messageQueue option when subscribing.
    public Subscription subscribe(String subject, String queueName, MessageQueue messageQueue);
}

class NatsConnection implements Connection {
    // The createSubscription method would receive the messageQueue param.
    NatsSubscription createSubscription(String subject, String queueName, NatsDispatcher dispatcher, MessageQueue messageQueue) {
        // ...
        sub = new NatsSubscription(sid, subject, queueName, this, dispatcher, messageQueue);
        // ...
    }
}

class NatsSubscription extends NatsConsumer implements Subscription {
    // The constructor would receive the message queue and set it.
    NatsSubscription(String sid, String subject, String queueName, NatsConnection connection,
                     NatsDispatcher dispatcher, MessageQueue incoming) {
        // ...
        this.dispatcher = dispatcher;
        this.incoming = incoming;
        this.usingExternalMessageQueue = false;

        if (this.incoming != null) {
            // I don't love this idea, but it could be a way of avoiding the subscription calling
            // "this.incoming.pause()" when "invalidate()" is invoked.
            this.usingExternalMessageQueue = true
        }

        // If both params are missing, we can manage our own.
        if (this.dispatcher == null && this.incoming == null) {
            this.incoming = new MessageQueue(false);
        }
    }

    void invalidate() {
        // Now we can use the new flag to avoid iterrupting an external message queue.
        if (!this.usingExternalMessageQueue && this.incoming != null) {
            this.incoming.pause();
        }
        this.dispatcher = null;
        // We should be able to set this to null since the subscription is invalid at this point.
        this.incoming = null;
    }
}
```

**Comment by sasbury** [comment]
I have been swamped but will try to grok this in the next week or so

**Comment by sasbury** [comment]
Would using multiple dispatchers be too heavy?

**Comment by film42** [comment]
@sasbury I think it would. Our implementation uses 10 subscriptions per nats client. Having 10 OS threads to replicate the behavior is really heavy.

**Comment by sasbury** [comment]
From slack - Maybe it makes sense to add a simple GroupSubscription abstraction? Something in-between a subscription and a dispatcher? Push the stats and network ops into a subscription group and then have the dispatcher use that under the hood but prevents duplicate subscriptions?

**Comment by sasbury** [comment]
Another slack comment - ultimately the goal is 1 subject -> 1 thread -> multiple subscribers

**Comment by sasbury** [comment]
released in 2.6.0
